### DISCUSSION TITLE
Is this code "bad" ? (Slower than php?)

### DISCUSSION THREAD

**Discussion by KevinBLT** [vlang/v#19553] [Questions and Answers] [answered]
Hello,

I implemented (more or less) the same code in V and PHP (a hypothetical log file test (5x50000 lines), searching for given passwords) and no matter what I did, it was at least a second slower than php. Doesn't seem correct, so I may have missed something?

(No real data involved!)

```v
import os

const (
	leaked_file = './leaked.txt'
)

fn main() {
	passwords := os.read_lines('./passwords.txt')!

	if os.exists(leaked_file) {
		os.rm(leaked_file)!
	}

	os.create(leaked_file)!

	mut out := os.open_append(leaked_file)!
 
	for f in os.ls('./')! {
		mut log := os.open(f)!

		if !f.contains('log.') {
			continue
		}

		for {
			mut buffer := []u8{len: 256}

			if log.read_bytes_into_newline(mut buffer)! > 0 {
		 
				for pw in passwords {
					line := buffer.bytestr()
 
					if line.contains(pw) {
						i1 := line.index(': ')? + 2
		 				i2 := line.index_after('failed', i1) - 1
 
						res := 'Leak found: ${line.substr(i1, i2)} | Password: ${pw} | Log File: ${f}'

						println(res)

						out.writeln(res)!
					}
				}
			} else {
				break
			}
		}

		log.close()
	}

	out.close()
}
```

PHP:
```php
<?php

define('LEAKED', './leaked.txt');

$passwords = explode("\n", file_get_contents('./passwords.txt'));

function read_lines($file_handle) { 
	while (!feof($file_handle)) {
		yield fgets($file_handle);
	}
}

if (file_exists(LEAKED)) {
	unlink(LEAKED);
}

file_put_contents(LEAKED, '');

if ($handle = opendir('.')) {
	while (false !== ($f = readdir($handle))) {
		if (str_contains($f, 'log.')) {
			foreach (read_lines(fopen($f, 'r')) as $line) {
				
				foreach ($passwords as $pw) {
					if (empty($pw)) continue;

					if (str_contains($line, $pw)) {
						$i1   = strpos($line, ': ') + 1;
						$i2   = strpos($line, 'failed') - 1;
						$name = substr($line, $i1, $i2 - $i1);
						$res  = "Leak found: $name | Password: $pw | Log File: $f";

						print $res . "\n";

						file_put_contents(LEAKED, "$res\n", FILE_APPEND);
					}
				}

			}
		}
	}

	closedir($handle);
}
```

V:  (Names are fake/generated)
<img width="682" alt="image" src="https://github.com/vlang/v/assets/12692409/4b874d60-f298-43cc-9d42-c4327b418df4">

PHP:
<img width="541" alt="image" src="https://github.com/vlang/v/assets/12692409/3cb34287-8437-41b2-9cab-4603790d788a">

I would assume, this it pretty much the same on C level?

Thanks for sharing information :)

**Comment by medvednikov** [comment]
have you tried using os.read_lines() for everything?

also compile with -prod?

**Reply by KevinBLT** [reply]
To not buffer every file into memory (each file is 5mb) I used filehandle.read_bytes_into_newline() and read until the end. The same way I did in php.

Yes I compliled with `v -prod -cc gcc -skip-unused -cflags -march=native v/main.v`. (It's visible in the screenshot).

Both codes are above. I would assume that they use the best way to read files line by line without holding everything in memory.
But nonetheless, I would assume that it has to be faster than php in any case but definitely not more than 1 second slower.

Both files load 4 strings from a file, list all files with `log.*` in current directory, read each file line by line (5 files, and 50000 lines each) and check if any line contains any of the strings.

**Reply by JalonSolov** [reply]
`-prod` forces gcc, so no need for the `-cc gcc`.  No need for `-cflags -march=native` either, as far as I'm aware, since it will build for the native system by default.

That said, it's impossible to test your code locally without having the files to read.  They don't have to be the originals (and I wouldn't want them anyway, since this is skirting illegalities), but something of the proper format is necessary.

If you can show the format of a line, a dummy file generator could likely be built fairly easily.

**Reply by KevinBLT** [reply]
Hello thanks for investigating!

It's pretty easy, you can clone this repo: https://github.com/koalasoft-de/find_it

As it is already generated fake data, no need to do anything else than cloning. 

! Maybe one additional info: I am using M1 Macbook - could this be an ARM issue?!?

The results are still the same with only `-prod`:
<img width="1132" alt="image" src="https://github.com/vlang/v/assets/12692409/8f70b360-773b-433d-8ef9-ca3f17723485">

**Reply by KevinBLT** [reply]
It's not an ARM thing, it's also slower on my slowest test system with an intel celeron cpu:
<img width="1129" alt="image" src="https://github.com/vlang/v/assets/12692409/77b6e067-53e8-484a-a6f2-949e6b4c722e">

**Reply by medvednikov** [reply]
I'll have a look, it shouldn't be slower.

will let you know

**Comment by lv37** [comment]
With this code its as fast as php on my machine at least

```v
module main

import os
import io

const leaked_file = './leaked.txt'

fn main() {
	passwords := os.read_lines('./passwords.txt')!
	if !os.exists(leaked_file) { os.create(leaked_file)! }

	mut out := io.new_multi_writer(os.stdout(), os.open_file(leaked_file, 'w')!)

	files := os.ls('.')!
	for f in files {
		if !f.starts_with('log.') { continue }

		mut log := io.new_buffered_reader(
			reader: os.open(f)!
		)
		for {
			line := log.read_line() or { break }
			for pw in passwords {
				if line.index_after(pw, 26) != -1 {
					res := 'Leak found: ${line[26..].before(' fail')} | Password: ${pw} | Log File: ${f}\n'
					out.write(res.bytes())!
					break
				}
			}
		}
	}
}
```

```bash
./main  0.26s user 0.02s system 96% cpu 0.282 total
```

**Reply by JalonSolov** [reply]
Yep.  In this case, it's more a matter of using a languages strengths, than trying to force it to act like another.

**Reply by KevinBLT** [reply]
Sorry, but I don't share that view. The code does nothing fancy and it's not trying to act as anything, just reading some lines into a buffer and checking for a substring. 

We are talking about more than one second and if this code is so significantly faster now, some function that was exchanged here has to be the one that is doing something slow and this should not be ignored. The alternative is that anyone trying v lang will just see results as I saw and will not consider this as fulfilling the claims of "the speed of c" (which took ~120 ms (on my machine of course)).

So from what I tested now is, that `read_bytes_into_newline` has to be slow somehow? 
A lot slower from what `io.new_buffered_reader(reader: os.open(f)!)` seems to be doing internally and I wouldn't consider this normal to be off by a more than a second from `read_bytes_into_newline()`.

**Reply by lv37** [reply]
Also replacing `line.contains(pw)` with `line.index_after(pw, 26) != -1` made it half a second faster

**Reply by lv37** [reply, marked as answer]
`str.contains(substr)` should also be optimized

**Reply by medvednikov** [reply]
Great job! Thanks, @wygsh

**Reply by Wajinn** [reply]
> We are talking about more than one second and if this code is so significantly faster now, some function that was exchanged here has to be the one that is doing something slow and this should not be ignored. The alternative is that anyone trying v lang will just see results as I saw and will not consider this as fulfilling the claims of "the speed of c" (which took ~120 ms (on my machine of course)).

I can understand where you are coming from, and it is good to point out things that might be an issue.  This will hopefully lead to performance improvements.  But, per JalonSolov, what he's saying has validity too.  How languages do things, don't always exactly line up, and when people are new and coming from a different language this can lead to some of them getting frustrated.  It's good to have some patience, research, or ask around (as you have done).

An example of that is this [discussion](https://github.com/vlang/v/issues/17698).  It started off with the claim V was ~145x slower.  Then some other programmers looked at the code, made some changes, and we ended it up seeing V was as fast and at times faster.  Of course each case can be different, and it can be that some improvements could be made.

**Reply by KevinBLT** [reply]
Yes that's true, you are right. My main feeling is, as it is still in beta or below 1.0, there can be some things that could be optimized. 
I am really interested in v lang as its style is almost everything I would like to have.

I have now tested around.
At first I though of cause `indexAfter(pw, 26)` is faster, because it does less, it doesn't check before index 26. 
But then I tried with 0 and it makes a huge difference again.

Look at this code:
```v
fn main() {
  text := '... LONG STRING...'
  
  mut iter := 1000000
  mut found := 0

  for {
    iter--

    if text.index_after('FLAG', 0) > -1 { // alternative: text.contains('FLAG')
      found++
    }

    if iter <= 0 {
      break
    }
  }

  println(found)

  exit(0)
}
```

Exchanging `index_after` and `contains` leads to more than 4 seconds difference. So from my point of view, just implement contains as `return text.indexAfter(search, 0) > -1` and it will have the same speed, otherwise what is the benefit of providing `str.contains` in the first place?

I am only doing some experiments here. I am pretty sure that someone will be stabbed in the back at some point in a larger program.
<img width="917" alt="image" src="https://github.com/vlang/v/assets/12692409/8d98bf1f-2603-44ea-834a-83c1e8f03f58">

I am neither a systems developer nor a php developer, the questions arose as I compared different programming languages with students in the test project and I found vlang before and thought about giving it a try. And as you can see in the screenshot, I am pretty sure this should not create such a gap. 

When I created the c code from the v code, I wondered why it's not using `strstr` from `<string.h>`, wouldn't that create leaner code? V ships its whole implementation, even stuff that's already there (and `<string.h>` is included). I don't quite understand all that.

I do understand that it may not be an always fair comparison but I still would say it should be possible to use a std lib function and expect the best possible implementation?

**Reply by JalonSolov** [reply]
The V stdlib has been made piecemeal over several years, with no concerted efforts to go back and run benchmarks, change older code to newer, etc.  It's definitely time to start doing those things, as V is about to hit 0.5 - the halfway mark to 1.0 release.

V is following the right way to do things, for now... "Make it work, first, _then_ optimize."

The main thing we're "waiting for" at the moment is a large rewrite being done on the parser, which will also affect other parts of the compiler.  The changes should make the compiler even faster, as well as enabling better code generation.
